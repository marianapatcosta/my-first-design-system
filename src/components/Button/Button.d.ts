import React, { CSSProperties, MouseEventHandler } from 'react';
declare type ButtonType = 'button' | 'submit' | 'reset' | undefined;
export declare type ButtonSize = 'small' | 'medium' | 'large';
export interface ButtonProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * Is this the principal call to action on the page?
     */
    primary?: boolean;
    /**
     * What is button's type
     */
    type?: ButtonType;
    /**
     * How large should the button be?
     */
    size?: ButtonSize;
    /**
     * Will the button occupy all the available width?
     */
    isFullLength?: boolean;
    /**
     * Button's label
     */
    label?: string;
    /**
     * Is Button disabled?
     */
    disabled?: boolean;
    /**
     * Button icon
     */
    icon?: string;
    /**
     * text to describe the button icon (for accessibility purposes)
     */
    iconText?: string;
    /**
     * Click handler
     */
    onClick: MouseEventHandler<HTMLButtonElement>;
}
/**
 * Primary UI component for user interaction
 */
export declare const Button: React.ForwardRefExoticComponent<ButtonProps & React.RefAttributes<HTMLButtonElement>>;
export {};
