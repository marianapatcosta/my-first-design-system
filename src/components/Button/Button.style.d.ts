import { ButtonSize } from './Button';
export declare const StyledButton: import("styled-components").StyledComponent<"button", any, {
    primary?: boolean | undefined;
    isFullLength?: boolean | undefined;
    size?: ButtonSize | undefined;
}, never>;
export declare const StyledButtonLabel: import("styled-components").StyledComponent<"span", any, {
    hasIcon?: boolean | undefined;
}, never>;
export declare const StyledButtonIcon: import("styled-components").StyledComponent<"img", any, {
    label?: string | undefined;
}, never>;
