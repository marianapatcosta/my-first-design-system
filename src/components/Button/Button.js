var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { forwardRef } from 'react';
import { StyledButton, StyledButtonLabel, StyledButtonIcon, } from './Button.style';
/**
 * Primary UI component for user interaction
 */
export var Button = forwardRef(function (_a, ref) {
    var label = _a.label, icon = _a.icon, iconText = _a.iconText, props = __rest(_a, ["label", "icon", "iconText"]);
    return (_jsxs(StyledButton, __assign({ ref: ref }, props, { children: [icon && (_jsx(StyledButtonIcon, { "data-testid": 'button-icon', src: icon, alt: iconText, label: label }, void 0)),
            label && (_jsx(StyledButtonLabel, __assign({ hasIcon: !!icon }, { children: label }), void 0))] }), void 0));
});
Button.defaultProps = {
    id: '',
    className: '',
    style: {},
    primary: false,
    size: 'medium',
    isFullLength: false,
    label: '',
    disabled: false,
    type: 'button',
    icon: '',
    iconText: '',
    onClick: function () { return null; }
};
