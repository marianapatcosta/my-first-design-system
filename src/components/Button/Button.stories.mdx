import { Meta } from '@storybook/addon-docs/blocks';
import { Button } from './Button';
import { Success } from '../../assets/icons';

<Meta
  title='Design System|Button'
  component={Button}
  argTypes={{ onClick: { action: 'clicked' } }}
/>

# Button

Simple button component

## Usage

Use button when it is required a click event to trigger a certain functionality

- Primary Button:

  <Story name='Primary'>
    <Button label='Button' primary />
  </Story>

- Secondary Button:

  <Story name='Secondary'>
    <Button label='Button' />
  </Story>

- FullLength Button:

  <Story name='FullLength'>
    <Button label='Full-length button' isFullLength />
  </Story>

- Small Button:

  <Story name='Small'>
    <Button label='Small button' size='small' />
  </Story>

- Medium Button:

  <Story name='Medium'>
    <Button label='Medium button' />
  </Story>

- Large Button:

  <Story name='Large'>
    <Button label='Large button' size='large' />
  </Story>

- Button Icon:

  <Story name='Large'>
    <Button icon={Success} primary />
  </Story>

- Button With icon:
  <Story name='Large'>
    <Button label='Button' size='large' icon={Success} primary />
  </Story>

## Other important notes

not applicable.
