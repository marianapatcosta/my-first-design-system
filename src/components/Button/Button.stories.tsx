import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Button, ButtonProps } from './Button';
import { Success } from '../../assets/icons';

export default {
  title: 'Button',
  component: Button,
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  primary: true,
  label: 'Primary button',
};

export const Secondary = Template.bind({});
Secondary.args = {
  label: 'Seconday button',
};

export const FullLength = Template.bind({});
FullLength.args = {
  isFullLength: true,
  label: 'Full length button',
};

export const Small = Template.bind({});
Small.args = {
  size: 'small',
  label: 'Small button',
};

export const Medium = Template.bind({});
Medium.args = {
  label: 'Medium button',
};

export const Large = Template.bind({});
Large.args = {
  size: 'large',
  label: 'This is large button',
};

export const Icon = Template.bind({});
Icon.args = {
  primary: true,
  icon: Success,
  size: 'small'
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  primary: true,
  label: 'icon button',
  icon: Success,
};
