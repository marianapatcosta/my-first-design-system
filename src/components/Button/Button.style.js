var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled, { css } from 'styled-components';
import { baseStyles, disabledStyles, truncateStyles, colors, } from '../../styles';
var sizeStyles = {
    small: css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    max-width: 4rem;\n    font-size: 90%;\n    padding: 0.3rem 0.625rem;\n  "], ["\n    max-width: 4rem;\n    font-size: 90%;\n    padding: 0.3rem 0.625rem;\n  "]))),
    medium: css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    max-width: 8rem;\n    font-size: 100%;\n    padding: 0.625rem 1.25rem;\n  "], ["\n    max-width: 8rem;\n    font-size: 100%;\n    padding: 0.625rem 1.25rem;\n  "]))),
    large: css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n    max-width: 12rem;\n    font-size: 100%;\n    padding: 0.625rem 1.25rem;\n  "], ["\n    max-width: 12rem;\n    font-size: 100%;\n    padding: 0.625rem 1.25rem;\n  "]))),
};
export var StyledButton = styled.button(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  ", ";\n  ", ";\n  display: flex;\n  align-content: center;\n  justify-content: center;\n  border: 0.063rem solid ", ";\n  border-radius: 0.2rem;\n  background-color: ", ";\n  color: ", ";\n  box-sizing: content-box;\n  cursor: pointer;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  :disabled,\n  :hover:disabled,\n  :active:disabled {\n    ", "\n  }\n\n  ", "\n\n  ", "\n"], ["\n  ", ";\n  ", ";\n  display: flex;\n  align-content: center;\n  justify-content: center;\n  border: 0.063rem solid ", ";\n  border-radius: 0.2rem;\n  background-color: ", ";\n  color: ", ";\n  box-sizing: content-box;\n  cursor: pointer;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  :disabled,\n  :hover:disabled,\n  :active:disabled {\n    ", "\n  }\n\n  ",
    "\n\n  ",
    "\n"])), baseStyles, function (_a) {
    var size = _a.size;
    return size && sizeStyles[size];
}, colors.highlight, colors.transparent, colors.highlight, disabledStyles, function (_a) {
    var primary = _a.primary;
    return primary &&
        "\n      color: " + colors.white + ";\n      background-color: " + colors.highlight + ";\n    ";
}, function (_a) {
    var isFullLength = _a.isFullLength;
    return isFullLength &&
        " max-width: 100%;\n      width: 100%;\n      font-size: 100%;\n\n    ";
});
export var StyledButtonLabel = styled.span(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  ", ";\n  user-select: none;\n\n  ", "\n"], ["\n  ", ";\n  user-select: none;\n\n  ",
    "\n"])), truncateStyles, function (_a) {
    var hasIcon = _a.hasIcon;
    return hasIcon &&
        "\n  line-height: 1rem;";
});
export var StyledButtonIcon = styled.img(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  width: 1rem;\n  user-select: none;\n\n  ", "\n"], ["\n  width: 1rem;\n  user-select: none;\n\n  ", "\n"])), function (_a) {
    var label = _a.label;
    return label && "margin-right: 0.5rem;";
});
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
