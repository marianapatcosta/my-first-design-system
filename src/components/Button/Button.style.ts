import styled from 'styled-components';
import { baseStyles, disabledStyles, colors } from '../../styles';

export const StyledButton = styled.button<{
  primary?: boolean;
  isFullLength?: boolean;
}>`
  ${baseStyles}
  padding: 0.625rem 1.25rem;
  border: 0.063rem solid ${colors.highlight};
  border-radius: 0.2rem;
  background-color: ${colors.transparent};
  color: ${colors.highlight};
  cursor: pointer;
  font-size: 90%;
  min-width: 6.25rem;

  :hover {
    opacity: 0.75;
  }

  :disabled,
  :hover:disabled,
  :active:disabled {
    ${disabledStyles}
  }

  ${({ primary }) =>
    primary &&
    `
    color: ${colors.white};
    background-color: ${colors.highlight};
    `}

  ${({ isFullLength }) =>
    isFullLength &&
    `
      width: 100%;
    `}
`;
