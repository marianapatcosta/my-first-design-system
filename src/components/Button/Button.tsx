import React, { CSSProperties, forwardRef, MouseEventHandler } from 'react';
import {
  StyledButton,
  StyledButtonLabel,
  StyledButtonIcon,
} from './Button.style';

type ButtonType = 'button' | 'submit' | 'reset' | undefined;
export type ButtonSize = 'small' | 'medium' | 'large';

export interface ButtonProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * Is this the principal call to action on the page?
   */
  primary?: boolean;
  /**
   * What is button's type
   */
  type?: ButtonType;
  /**
   * How large should the button be?
   */
  size?: ButtonSize;
  /**
   * Will the button occupy all the available width?
   */
  isFullLength?: boolean;
  /**
   * Button's label
   */
  label?: string;
  /**
   * Is Button disabled?
   */
  disabled?: boolean;
  /**
   * Button icon
   */
  icon?: string;
  /**
   * text to describe the button icon (for accessibility purposes)
   */
  iconText?: string;
  /**
   * Click handler
   */
  onClick: MouseEventHandler<HTMLButtonElement>;
}

/**
 * Primary UI component for user interaction
 */
export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ label, icon, iconText, ...props }, ref) => {
    return (
      <StyledButton ref={ref} {...props}>
        {icon && (
          <StyledButtonIcon
            data-testid='button-icon'
            src={icon}
            alt={iconText}
            label={label}
          />
        )}
        {label && (
          <StyledButtonLabel hasIcon={!!icon}>{label}</StyledButtonLabel>
        )}
      </StyledButton>
    );
  }
);

Button.defaultProps = {
  id: '',
  className: '',
  style: {},
  primary: false,
  size: 'medium',
  isFullLength: false,
  label: '',
  disabled: false,
  type: 'button',
  icon: '',
  iconText: '',
};
