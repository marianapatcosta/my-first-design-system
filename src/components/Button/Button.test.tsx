import React from 'react';
import { render, getByText, fireEvent } from '@testing-library/react';
import { Button } from './Button';
import { Success } from '../../assets/icons';
import { colors } from '../../styles';
import { hexToRGB } from '../../utils/utils';

describe('Button', () => {
  const defaultProps = {
    label: 'Button',
    onClick: () => null,
  };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Button {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display label', () => {
    const { container } = render(<Button {...defaultProps} />);
    getByText(container, 'Button');
  });

  test('should handle click events', () => {
    const onClickMock = jest.fn();
    const {
      container: { firstChild },
    } = render(<Button {...defaultProps} onClick={onClickMock} />);

    firstChild && fireEvent.click(firstChild);

    expect(onClickMock).toBeCalled();
  });

  test('should return null on click if onClick prop is not passed', () => {
    const result = Button.defaultProps?.onClick && Button.defaultProps.onClick();
    expect(result).toBeNull();
  });

  describe('Primary button', () => {
    test('should be rendered', () => {
      const { getByRole } = render(<Button {...defaultProps} primary />);
      const button = getByRole('button');
      const style = window.getComputedStyle(button);
      expect(style.backgroundColor).toBe(hexToRGB(colors.highlight));
    });
  });

  describe('Full length button', () => {
    test('should be rendered', () => {
      const { getByRole } = render(<Button {...defaultProps} isFullLength />);
      const button = getByRole('button');
      const style = window.getComputedStyle(button);
      expect(style.maxWidth).toBe('100%');
    });
  });

  describe('Button with icon', () => {
    test('should render icon if icon prop is passed', () => {
      const { getByTestId } = render(
        <Button {...defaultProps} icon={Success} />
      );
      const icon = getByTestId('button-icon');
      expect(icon).toBeTruthy();
      expect(icon).toHaveAttribute('src', Success);
    });
  });

  describe('Disabled Button', () => {
    test('should display text', () => {
      const { container } = render(
        <Button {...defaultProps} disabled={true} />
      );
      getByText(container, 'Button');
    });

    test('should be disabled', () => {
      const {
        container: { firstChild },
      } = render(<Button {...defaultProps} disabled={true} />);

      // expect(firstChild).toBeDisabled();
      expect(firstChild).toHaveAttribute('disabled');
    });

    test('should not handle click events', () => {
      const onClickMock = jest.fn();
      const {
        container: { firstChild },
      } = render(<Button {...defaultProps} disabled onClick={onClickMock} />);

      // fireEvent.click(component);
      firstChild && fireEvent.click(firstChild);

      expect(onClickMock).not.toBeCalled();
    });
  });
});
