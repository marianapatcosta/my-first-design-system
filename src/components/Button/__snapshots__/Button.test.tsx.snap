// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders without errors and matches snapshot 1`] = `
<button
  class="sc-bdfBwQ fgPrFZ"
  id=""
  type="button"
>
  
  <span
    class="sc-gsTCUz gffxlH"
  >
    Button
  </span>
</button>
`;
